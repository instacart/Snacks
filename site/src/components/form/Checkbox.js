import React, {Fragment} from 'react'
import Playground from '../Playground'
import {Title3} from '../Typography'

export default () => (
  <Fragment>
    This component outputs a checkbox with an optional label.

    You can define whether the button starts in the pressed on unpressed state through the `isSelected` prop. In addition, you can change the selected/unselected state at any moment after the mount state by redifining the prop `isSelected`.

    <Title3>Label</Title3>
    Associating a label to every radio button is very important for accessibility purposes. There are two ways to do it. The easiest is to enclose the text with the component: <code>{`<Checkbox>My label</Checkbox>`}</code>. The other option is to associate a label yourself:

    <Playground>
      {`
        <div style={{display: 'flex', flexDirection: 'column'}}>
          <div style={{display: 'flex', marginBottom: 20}}>
            <Checkbox id="checkbox1" isSelected />
            <label htmlFor="checkbox1" style={{marginLeft: 10}}>Label associated manually</label>
          </div>
          <Checkbox id="checkbox2">Label auto-generated</Checkbox>
        </div>
      `}
    </Playground>

    <Title3>Styling</Title3>
    You can pass an object to the `style` prop for styling the label, radio button and wrap element using Radium's structure:

    <Playground>
      {`
        <Checkbox
          id="checkbox3"
          wrapEl='span'
          style={{
            wrapEl: {padding: 10, border: '2px dotted lightblue'},
            button: {marginRight: 40},
            label: {color: 'salmon'},
          }}
        >
          My Checkbox Button Label
        </Checkbox>
      `}
    </Playground>


    <Title3>Callback</Title3>
    You can pass a callback through the `onChange` prop:

    <Playground>
      {`
        function onChange(event, props) {
          alert(\`This button is \${props.isSelected ? '' : 'not '}selected\`)
        }

        <Checkbox id="checkbox4" onChange={onChange}>Click me!</Checkbox>
      `}
    </Playground>

    <Title3>Disabling</Title3>
    You can disable a checkbox by declaring the `isDisabled` prop:

    <Playground>
      {`
        function onChange(event, props) {
          alert('This function will not be called')
        }

        <Checkbox id  ="checkbox5" onChange={onChange} isDisabled>I am disabled</Checkbox>
      `}
    </Playground>
    **Important!** If you attach a label to a checkbox manually, make sure the colors for the disabled state follow the styleguide.
  </Fragment>
)
