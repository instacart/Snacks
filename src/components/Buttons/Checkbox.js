import React from 'react'
import PropTypes from 'prop-types'
import RadioCheckboxBase from '../../base/RadioCheckboxBase'
import checkboxBaseSvg from '../../assets/checkboxBase.svg'
import checkboxSelectedSvg from '../../assets/checkboxSelected.svg'

const renderInputButton = (isSelected, style) => {
  const SvgComponent = isSelected ? checkboxSelectedSvg : checkboxBaseSvg
  return <SvgComponent aria-hidden="true" style={style}/>
}

const Checkbox = (props) => {
  return (
    <RadioCheckboxBase
      btnType='checkbox'
      renderInputButton={renderInputButton}
      {...props}
    />
  )
}

Checkbox.propTypes = {
  aria          : PropTypes.shape({
    label         :PropTypes.string,
  }),
  children      : PropTypes.node,
  id            : PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  isDisabled    : PropTypes.bool,
  isSelected    : PropTypes.bool,
  name          : PropTypes.string,
  onChange      : PropTypes.func,
  onBlur        : PropTypes.func,
  onFocus       : PropTypes.func,
  style         : PropTypes.shape({
    button        : PropTypes.object,
    label         : PropTypes.object,
    wrapEl        : PropTypes.object,
  }),
  value         : PropTypes.string,
  wrapEl        : PropTypes.string,
}

export default Checkbox
