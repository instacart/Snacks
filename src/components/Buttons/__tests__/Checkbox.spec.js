import Checkbox from '../Checkbox'
import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import sinon from 'sinon'

describe('Checkbox', () => {
  it('renders the correct button type', () => {
    const tree = renderer
      .create(<Checkbox id={1} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("renders the correct selected state", () => {
    const testCases = [
      { id: 1, isSelected: false },
      { id: 2, isSelected: true },
      { id: 2, isSelected: true, isDisabled: true },
    ]

    testCases.forEach(props => {
      const tree = renderer
        .create(<Checkbox {...props} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('sets isSelected to false if isDisabled is set to true', () => {
    const wrapper = mount(<Checkbox id="A" value="test" isSelected>Testing disabled</Checkbox>)
    const htmlBtn = wrapper.find('input').instance()

    expect(htmlBtn.checked).toBe(true)
    expect(htmlBtn.disabled).toBe(false)
    
    wrapper.setProps({isDisabled: true})

    expect(htmlBtn.checked).toBe(false)
    expect(htmlBtn.disabled).toBe(true)
  })

  
  it('generates a label if a child prop of text is supplied', () => {
    const tree = renderer
      .create(<Checkbox id={1}>Test label</Checkbox>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  });
  

  it('incorporates user style if passed', () => {
    const testCases = [
      {
        id: 1,
        style: {
          wrapEl: {background: 'pink'},
          button: {border: '2px solid purple' },
          label: {color: 'green'}
        },
      },
      {
        id: 2,
        style: undefined
      },
    ]

    testCases.forEach(props => {
      const tree = renderer
        .create(<Checkbox {...props}>Test label</Checkbox>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('calls the user-supplied onChange function', () => {
    const onChange = sinon.spy()
    const wrapper = mount(<Checkbox id={1} onChange={onChange} />)

    wrapper.find('input').simulate('change')
    expect(onChange.calledOnce).toBe(true)
  })
})
