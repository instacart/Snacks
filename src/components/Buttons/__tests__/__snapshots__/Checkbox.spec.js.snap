// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox generates a label if a child prop of text is supplied 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <test-file-stub
      aria-hidden="true"
      data-radium={true}
      style={
        Object {
          "fill": "#757575",
          "height": 22,
          "width": 22,
        }
      }
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      disabled={undefined}
      id={1}
      onBlur={undefined}
      onChange={[Function]}
      onFocus={undefined}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "backgroundImage": "none",
          "height": 22,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="checkbox"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={1}
    style={
      Object {
        "lineHeight": "22px",
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Checkbox incorporates user style if passed 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "background": "pink",
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "border": "2px solid purple",
        "position": "relative",
      }
    }
  >
    <test-file-stub
      aria-hidden="true"
      data-radium={true}
      style={
        Object {
          "fill": "#757575",
          "height": 22,
          "width": 22,
        }
      }
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      disabled={undefined}
      id={1}
      onBlur={undefined}
      onChange={[Function]}
      onFocus={undefined}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "backgroundImage": "none",
          "height": 22,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="checkbox"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={1}
    style={
      Object {
        "color": "green",
        "lineHeight": "22px",
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Checkbox incorporates user style if passed 2`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <test-file-stub
      aria-hidden="true"
      data-radium={true}
      style={
        Object {
          "fill": "#757575",
          "height": 22,
          "width": 22,
        }
      }
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      disabled={undefined}
      id={2}
      onBlur={undefined}
      onChange={[Function]}
      onFocus={undefined}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "backgroundImage": "none",
          "height": 22,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="checkbox"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={2}
    style={
      Object {
        "lineHeight": "22px",
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Checkbox renders the correct button type 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <test-file-stub
    aria-hidden="true"
    data-radium={true}
    style={
      Object {
        "fill": "#757575",
        "height": 22,
        "width": 22,
      }
    }
  />
  <input
    aria-label={undefined}
    checked={false}
    data-radium={true}
    disabled={undefined}
    id={1}
    onBlur={undefined}
    onChange={[Function]}
    onFocus={undefined}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;

exports[`Checkbox renders the correct selected state 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <test-file-stub
    aria-hidden="true"
    data-radium={true}
    style={
      Object {
        "fill": "#757575",
        "height": 22,
        "width": 22,
      }
    }
  />
  <input
    aria-label={undefined}
    checked={false}
    data-radium={true}
    disabled={undefined}
    id={1}
    onBlur={undefined}
    onChange={[Function]}
    onFocus={undefined}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;

exports[`Checkbox renders the correct selected state 2`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <test-file-stub
    aria-hidden="true"
    data-radium={true}
    style={
      Object {
        "fill": "#43B02A",
        "height": 22,
        "width": 22,
      }
    }
  />
  <input
    aria-label={undefined}
    checked={true}
    data-radium={true}
    disabled={undefined}
    id={2}
    onBlur={undefined}
    onChange={[Function]}
    onFocus={undefined}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;

exports[`Checkbox renders the correct selected state 3`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <test-file-stub
    aria-hidden="true"
    data-radium={true}
    style={
      Object {
        "fill": "#BDBDBD",
        "height": 22,
        "width": 22,
      }
    }
  />
  <input
    aria-label={undefined}
    checked={true}
    data-radium={true}
    disabled={true}
    id={2}
    onBlur={undefined}
    onChange={[Function]}
    onFocus={undefined}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "backgroundImage": "none",
        "height": 22,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;
